corehq/form_processor/models.py

```diff
@@ -857,6 +862,20 @@ def _saved_indices(self):
 
         return CaseAccessorSQL.get_indices(self.domain, self.case_id) if self.is_saved() else []
 
+    def _set_indices(self, value):
+        """Set previously-saved indices
+
+        Private setter used by the class constructor to populate indices
+        from a source such as ElasticSearch. This setter does not update
+        tracked models, and therefore is not intended for use with cases
+        whose state is being mutated.
+
+        :param value: A list of dicts that will be used to construct
+        `CommCareCaseIndexSQL` objects.
+        """
+        cache = self._saved_indices.get_cache(self)
+        cache[()] = [CommCareCaseIndexSQL(**x) for x in value]
```

Comment:
The `_saved_indices` method also supports returning the `cached_indices` attribute which is used in other places in the code to achieve a similar result.

I think it would be good to have a consistent method for setting these related models.

The method you use here also relies on `_saved_indices` having the `memoized` decorator and the link between the two is not obvious (particularly if you're looking at the `_saved_indices` function). It also requires understanding how memoized constructs cache keys which is somewhat of an internal implementation detail.

Reply:
I'm not sure why I overlooked `cached_indices`, although it feels a little weird to have two different kinds of caching in `_saved_indices`. f3666645b49df4047b5095e0741e7c989eef5426

Reply:
I agree that it's not ideal to have two methods. I think I added the `cached_indices` mechanism because it was simpler than using the internals of memoized but we could change how it works. I think it's only this one and `attachments_list` that would need to change (both used by `_attach_prefetch_models`)